File: golang/tests/pkg/integration/context_sync_mapping_policy_test.go
package integration

import (
	"fmt"
	"os"
	"strings"
	"testing"
	"time"

	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// TestContextSyncWithMappingPolicyIntegration tests cross-platform context synchronization with MappingPolicy
// This test simulates: Windows session -> persist data -> Linux session -> access data
func TestContextSyncWithMappingPolicyIntegration(t *testing.T) {
	// Skip in CI environment or if API key is not available
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" || os.Getenv("CI") != "" {
		t.Skip("Skipping integration test: No API key available or running in CI")
	}

	// Initialize the AgentBay client
	ab, err := agentbay.NewAgentBay(apiKey)
	require.NoError(t, err, "Failed to create AgentBay client")

	// 1. Create a unique context name
	contextName := fmt.Sprintf("test-mapping-policy-%d", time.Now().Unix())
	contextResult, err := ab.Context.Get(contextName, true)
	require.NoError(t, err, "Error getting/creating context")
	require.NotNil(t, contextResult.Context, "Context should not be nil")

	context := contextResult.Context
	t.Logf("Created context: %s (ID: %s)", context.Name, context.ID)

	// Ensure context is deleted after the test
	defer func() {
		deleteContextResult, err := ab.Context.Delete(context)
		if err != nil {
			t.Logf("Warning: Failed to delete context: %v", err)
		} else {
			t.Logf("Context deleted: %s (RequestID: %s)", context.ID, deleteContextResult.RequestID)
		}
	}()

	// Define paths
	windowsPath := "c:\\Users\\Administrator\\Downloads"
	linuxPath := "/home/wuying/下载"
	testFileName := "cross-platform-test.txt"
	testContent := "This file was created in Windows session and should be accessible in Linux session"

	// ========== Phase 1: Create Windows session and persist data ==========
	t.Log("========== Phase 1: Windows Session - Create and Persist Data ==========")

	// Create sync policy for Windows session (no mapping policy needed for first session)
	windowsSyncPolicy := &agentbay.SyncPolicy{
		UploadPolicy:   agentbay.NewUploadPolicy(),
		DownloadPolicy: agentbay.NewDownloadPolicy(),
		DeletePolicy:   agentbay.NewDeletePolicy(),
		ExtractPolicy:  agentbay.NewExtractPolicy(),
	}

	// Create Windows session with context sync
	windowsSessionParams := agentbay.NewCreateSessionParams()
	windowsSessionParams.AddContextSync(context.ID, windowsPath, windowsSyncPolicy)
	windowsSessionParams.WithImageId("windows_latest")
	windowsSessionParams.WithLabels(map[string]string{
		"test": "mapping-policy-windows",
	})

	// Create Windows session
	windowsSessionResult, err := ab.Create(windowsSessionParams)
	require.NoError(t, err, "Error creating Windows session")
	require.NotNil(t, windowsSessionResult.Session, "Windows session should not be nil")

	windowsSession := windowsSessionResult.Session
	t.Logf("Created Windows session: %s", windowsSession.SessionID)

	// Wait for Windows session to be ready
	t.Log("Waiting for Windows session to be ready...")
	time.Sleep(15 * time.Second)

	// Create directory in Windows session
	t.Logf("Creating directory in Windows: %s", windowsPath)
	windowsDirResult, err := windowsSession.FileSystem.CreateDirectory(windowsPath)
	require.NoError(t, err, "Error creating Windows directory")
	assert.NotEmpty(t, windowsDirResult.RequestID, "Windows directory creation request ID should not be empty")

	// Create test file in Windows session
	testFilePath := windowsPath + "\\" + testFileName
	t.Logf("Creating test file in Windows: %s", testFilePath)
	createFileCmd := fmt.Sprintf("echo %s > \"%s\"", testContent, testFilePath)
	windowsCmdResult, err := windowsSession.Command.ExecuteCommand(createFileCmd)
	require.NoError(t, err, "Error creating test file in Windows")
	t.Logf("Windows file creation result: %+v", windowsCmdResult)

	// Verify file exists in Windows session
	verifyFileCmd := fmt.Sprintf("type \"%s\"", testFilePath)
	verifyResult, err := windowsSession.Command.ExecuteCommand(verifyFileCmd)
	require.NoError(t, err, "Error verifying file in Windows")
	t.Logf("Windows file content: %s", verifyResult.Output)
	assert.Contains(t, verifyResult.Output, testContent, "File should contain test content in Windows")

	// Sync Windows session to upload data
	t.Log("Syncing Windows session to upload data...")
	windowsSyncResult, err := windowsSession.Context.Sync()
	require.NoError(t, err, "Error syncing Windows context")
	require.True(t, windowsSyncResult.Success, "Windows context sync should be successful")
	t.Logf("Windows context sync successful (RequestID: %s)", windowsSyncResult.RequestID)

	// Wait for upload to complete
	t.Log("Waiting for upload to complete...")
	time.Sleep(10 * time.Second)

	// Delete Windows session
	t.Log("Deleting Windows session...")
	windowsDeleteResult, err := ab.Delete(windowsSession, true)
	require.NoError(t, err, "Error deleting Windows session")
	t.Logf("Windows session deleted: %s (RequestID: %s)", windowsSession.SessionID, windowsDeleteResult.RequestID)

	// ========== Phase 2: Create Linux session with MappingPolicy and verify data ==========
	t.Log("========== Phase 2: Linux Session - Access Data via MappingPolicy ==========")

	// Create mapping policy with Windows path
	mappingPolicy := &agentbay.MappingPolicy{
		Path: windowsPath,
	}

	// Create sync policy with mapping policy for Linux session
	linuxSyncPolicy := &agentbay.SyncPolicy{
		UploadPolicy:   agentbay.NewUploadPolicy(),
		DownloadPolicy: agentbay.NewDownloadPolicy(),
		DeletePolicy:   agentbay.NewDeletePolicy(),
		ExtractPolicy:  agentbay.NewExtractPolicy(),
		MappingPolicy:  mappingPolicy,
	}

	// Create Linux session with context sync and mapping policy
	linuxSessionParams := agentbay.NewCreateSessionParams()
	linuxSessionParams.AddContextSync(context.ID, linuxPath, linuxSyncPolicy)
	linuxSessionParams.WithImageId("linux_latest")
	linuxSessionParams.WithLabels(map[string]string{
		"test": "mapping-policy-linux",
	})

	// Create Linux session
	linuxSessionResult, err := ab.Create(linuxSessionParams)
	require.NoError(t, err, "Error creating Linux session")
	require.NotNil(t, linuxSessionResult.Session, "Linux session should not be nil")

	linuxSession := linuxSessionResult.Session
	t.Logf("Created Linux session: %s with mapping from %s to %s", linuxSession.SessionID, windowsPath, linuxPath)

	// Ensure Linux session is deleted after the test
	defer func() {
		deleteResult, err := ab.Delete(linuxSession, true)
		if err != nil {
			t.Logf("Warning: Failed to delete Linux session: %v", err)
		} else {
			t.Logf("Linux session deleted: %s (RequestID: %s)", linuxSession.SessionID, deleteResult.RequestID)
		}
	}()

	// Wait for Linux session to be ready and data to be downloaded
	t.Log("Waiting for Linux session to be ready and data to be downloaded...")
	time.Sleep(15 * time.Second)

	// Verify file exists in Linux session at the mapped path
	linuxTestFilePath := linuxPath + "/" + testFileName
	t.Logf("Verifying file exists in Linux at: %s", linuxTestFilePath)

	// Check if file exists
	checkFileCmd := fmt.Sprintf("test -f \"%s\" && echo 'FILE_EXISTS' || echo 'FILE_NOT_FOUND'", linuxTestFilePath)
	checkResult, err := linuxSession.Command.ExecuteCommand(checkFileCmd)
	require.NoError(t, err, "Error checking file existence in Linux")
	t.Logf("Linux file check result: %s", checkResult.Output)

	// Verify file exists
	assert.Contains(t, checkResult.Output, "FILE_EXISTS", "File should exist in Linux session at mapped path")

	// Read file content in Linux session
	readFileCmd := fmt.Sprintf("cat \"%s\"", linuxTestFilePath)
	readResult, err := linuxSession.Command.ExecuteCommand(readFileCmd)
	require.NoError(t, err, "Error reading file in Linux")
	t.Logf("Linux file content: %s", readResult.Output)

	// Verify file content matches
	assert.True(t, strings.Contains(readResult.Output, testContent) || strings.Contains(readResult.Output, strings.TrimSpace(testContent)),
		"File content in Linux should match the content created in Windows")

	// Verify context info
	contextInfo, err := linuxSession.Context.Info()
	require.NoError(t, err, "Error getting context info")
	assert.NotEmpty(t, contextInfo.RequestID, "Request ID should not be empty")

	if len(contextInfo.ContextStatusData) > 0 {
		t.Log("Context status data in Linux session:")
		for i, data := range contextInfo.ContextStatusData {
			t.Logf("Context Status Data [%d]:", i)
			t.Logf("  ContextId: %s", data.ContextId)
			t.Logf("  Path: %s", data.Path)
			t.Logf("  Status: %s", data.Status)
			t.Logf("  TaskType: %s", data.TaskType)
		}

		// Verify the context data
		for _, data := range contextInfo.ContextStatusData {
			if data.ContextId == context.ID {
				assert.Equal(t, linuxPath, data.Path, "Path should match the Linux path")
			}
		}
	}

	t.Log("========== Cross-platform mapping policy test completed successfully ==========")
	t.Log("✓ Data created in Windows session was successfully accessed in Linux session via MappingPolicy")
}

File: typescript/tests/integration/context-sync-mapping-policy.test.ts
import { AgentBay } from "../../src/agent-bay";
import {
  newUploadPolicy,
  newDownloadPolicy,
  newDeletePolicy,
  newExtractPolicy,
  MappingPolicy,
  SyncPolicy,
} from "../../src/context-sync";
import { CreateSessionParams } from "../../src/session-params";

describe("Context Sync with MappingPolicy Integration Tests", () => {
  let agentBay: AgentBay;
  const apiKey = process.env.AGENTBAY_API_KEY;

  beforeAll(() => {
    if (!apiKey || process.env.CI) {
      console.log("Skipping integration test: No API key available or running in CI");
      return;
    }
    agentBay = new AgentBay({ apiKey });
  });

  it("should create session with MappingPolicy for cross-platform persistence", async () => {
    if (!apiKey || process.env.CI) {
      console.log("Skipping test: No API key available or running in CI");
      return;
    }

    // 1. Create a unique context name
    const contextName = `test-mapping-policy-${Date.now()}`;
    const contextResult = await agentBay.context.get(contextName, true);
    expect(contextResult.context).toBeDefined();

    const context = contextResult.context!;
    console.log(`Created context: ${context.name} (ID: ${context.id})`);

    try {
      // Define paths
      const windowsPath = "c:\\Users\\Administrator\\Downloads";
      const linuxPath = "/home/wuying/下载";
      const testFileName = "cross-platform-test.txt";
      const testContent = "This file was created in Windows session and should be accessible in Linux session";

      // ========== Phase 1: Create Windows session and persist data ==========
      console.log("========== Phase 1: Windows Session - Create and Persist Data ==========");

      // Create sync policy for Windows session (no mapping policy needed for first session)
      const windowsSyncPolicy: SyncPolicy = {
        uploadPolicy: newUploadPolicy(),
        downloadPolicy: newDownloadPolicy(),
        deletePolicy: newDeletePolicy(),
        extractPolicy: newExtractPolicy(),
      };

      // Create Windows session with context sync
      const windowsSessionParams = new CreateSessionParams();
      windowsSessionParams.addContextSync(context.id, windowsPath, windowsSyncPolicy);
      windowsSessionParams.withImageId("windows_latest");
      windowsSessionParams.withLabels({ test: "mapping-policy-windows" });

      // Create Windows session
      const windowsSessionResult = await agentBay.create(windowsSessionParams);
      expect(windowsSessionResult.session).toBeDefined();

      const windowsSession = windowsSessionResult.session!;
      console.log(`Created Windows session: ${windowsSession.sessionId}`);

      // Wait for Windows session to be ready
      console.log("Waiting for Windows session to be ready...");
      await new Promise((resolve) => setTimeout(resolve, 15000));

      // Create directory in Windows session
      console.log(`Creating directory in Windows: ${windowsPath}`);
      const windowsDirResult = await windowsSession.fileSystem.createDirectory(windowsPath);
      expect(windowsDirResult.requestId).toBeDefined();

      // Create test file in Windows session
      const testFilePath = `${windowsPath}\\${testFileName}`;
      console.log(`Creating test file in Windows: ${testFilePath}`);
      const createFileCmd = `echo ${testContent} > "${testFilePath}"`;
      const windowsCmdResult = await windowsSession.command.executeCommand(createFileCmd);
      expect(windowsCmdResult.output).toBeDefined();
      console.log(`Windows file creation result: ${windowsCmdResult}`);

      // Verify file exists in Windows session
      const verifyFileCmd = `type "${testFilePath}"`;
      const verifyResult = await windowsSession.command.executeCommand(verifyFileCmd);
      expect(verifyResult.output).toBeDefined();
      console.log(`Windows file content: ${verifyResult.output}`);
      expect(verifyResult.output).toContain(testContent);

      // Sync Windows session to upload data
      console.log("Syncing Windows session to upload data...");
      const windowsSyncResult = await windowsSession.context.sync();
      expect(windowsSyncResult.success).toBe(true);
      console.log(`Windows context sync successful (RequestID: ${windowsSyncResult.requestId})`);

      // Wait for upload to complete
      console.log("Waiting for upload to complete...");
      await new Promise((resolve) => setTimeout(resolve, 10000));

      // Delete Windows session
      console.log("Deleting Windows session...");
      const windowsDeleteResult = await agentBay.delete(windowsSession, true);
      console.log(
        `Windows session deleted: ${windowsSession.sessionId} (RequestID: ${windowsDeleteResult.requestId})`
      );

      // ========== Phase 2: Create Linux session with MappingPolicy and verify data ==========
      console.log("========== Phase 2: Linux Session - Access Data via MappingPolicy ==========");

      // Create mapping policy with Windows path
      const mappingPolicy: MappingPolicy = {
        path: windowsPath,
      };

      // Create sync policy with mapping policy for Linux session
      const linuxSyncPolicy: SyncPolicy = {
        uploadPolicy: newUploadPolicy(),
        downloadPolicy: newDownloadPolicy(),
        deletePolicy: newDeletePolicy(),
        extractPolicy: newExtractPolicy(),
        mappingPolicy: mappingPolicy,
      };

      // Create Linux session with context sync and mapping policy
      const linuxSessionParams = new CreateSessionParams();
      linuxSessionParams.addContextSync(context.id, linuxPath, linuxSyncPolicy);
      linuxSessionParams.withImageId("linux_latest");
      linuxSessionParams.withLabels({ test: "mapping-policy-linux" });

      // Create Linux session
      const linuxSessionResult = await agentBay.create(linuxSessionParams);
      expect(linuxSessionResult.session).toBeDefined();

      const linuxSession = linuxSessionResult.session!;
      console.log(
        `Created Linux session: ${linuxSession.sessionId} with mapping from ${windowsPath} to ${linuxPath}`
      );

      try {
        // Wait for Linux session to be ready and data to be downloaded
        console.log("Waiting for Linux session to be ready and data to be downloaded...");
        await new Promise((resolve) => setTimeout(resolve, 15000));

        // Verify file exists in Linux session at the mapped path
        const linuxTestFilePath = `${linuxPath}/${testFileName}`;
        console.log(`Verifying file exists in Linux at: ${linuxTestFilePath}`);

        // Check if file exists
        const checkFileCmd = `test -f "${linuxTestFilePath}" && echo "FILE_EXISTS" || echo "FILE_NOT_FOUND"`;
        const checkResult = await linuxSession.command.executeCommand(checkFileCmd);
        expect(checkResult).toBeDefined();
        console.log(`Linux file check result: ${checkResult.output}`);

        // Verify file exists
        expect(checkResult.output).toContain("FILE_EXISTS");

        // Read file content in Linux session
        const readFileCmd = `cat "${linuxTestFilePath}"`;
        const readResult = await linuxSession.command.executeCommand(readFileCmd);
        expect(readResult).toBeDefined();
        console.log(`Linux file content: ${readResult.output}`);

        // Verify file content matches
        expect(
          readResult.output.includes(testContent) || readResult.output.includes(testContent.trim())
        ).toBe(true);

        // Verify context info
        const contextInfo = await linuxSession.context.info();
        expect(contextInfo.requestId).toBeDefined();

        if (contextInfo.contextStatusData && contextInfo.contextStatusData.length > 0) {
          console.log("Context status data in Linux session:");
          contextInfo.contextStatusData.forEach((data: any, i: number) => {
            console.log(`Context Status Data [${i}]:`);
            console.log(`  ContextId: ${data.contextId}`);
            console.log(`  Path: ${data.path}`);
            console.log(`  Status: ${data.status}`);
            console.log(`  TaskType: ${data.taskType}`);
          });

          // Verify the context data
          const contextData = contextInfo.contextStatusData.find(
            (data: any) => data.contextId === context.id
          );
          if (contextData) {
            expect(contextData.path).toBe(linuxPath);
          }
        }

        console.log("========== Cross-platform mapping policy test completed successfully ==========");
        console.log("✓ Data created in Windows session was successfully accessed in Linux session via MappingPolicy");
      } finally {
        // Ensure Linux session is deleted
        const deleteResult = await agentBay.delete(linuxSession, true);
        console.log(
          `Linux session deleted: ${linuxSession.sessionId} (RequestID: ${deleteResult.requestId})`
        );
      }
    } finally {
      // Ensure context is deleted
      const deleteContextResult = await agentBay.context.delete(context);
      console.log(
        `Context deleted: ${context.id} (RequestID: ${deleteContextResult.requestId})`
      );
    }
  }, 180000);
});
